### ISSUE TITLE
Replace `AlgorithmResult` with `NodeState`

### ISSUE THREAD

**Issue by ljeub-pometry** [Pometry/Raphtory#1426]
- `group_by` should return groups as lists of nodes instead of transforming to string names
- would be nicer if groups were returned as our own object instead of a hash map so we can have additional methods such as `.as_subgraphs`? 
- the internal HashMap uses `usize` instead of `VID` which complicates things for no reason
- `get_all` seems useless?

**Comment by miratepuffin** [comment]
Adding other tickets which can go in here - #910

**Comment by ljeub-pometry** [comment]
The group_by suggestions are not implemented yet and need to be kept in mind in the move to `NodeState`

**Comment by wyatt-joyner-pometry** [comment]
This is all going to be subsumed by converting outputs to NodeState, right,  @ljeub-pometry?

**Comment by ljeub-pometry** [comment]
Indeed, that’s the plan. The ‘group_by’ related things still need to be implemented for ‘NodeState’ though
